### PR TITLE
planner: refactor a few code of plan cache

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #54057

Problem Summary: planner: refactor a few code of plan cache

### What changed and how does it work?

Add table dirty flags into the plan cache key

No regression

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/plan_cache_test.go [hunk 1]
```diff
@@ -1837,3 +1837,48 @@ func TestIndexRange(t *testing.T) {
...
+					}
+					if testT2Dirty {
+						tk.MustExec(`insert into t2 values (1)`)
+					}
+					tk.MustExec(`execute st`)
+
+					if testT1Dirty == t1Dirty && testT2Dirty == t2Dirty {
+						tk.MustQuery(`select @@last_plan_from_cache`).Check(testkit.Rows("1"))
+					} else {
+						tk.MustQuery(`select @@last_plan_from_cache`).Check(testkit.Rows("1"))
```
Comment (AilinKid) [inline pkg/planner/core/plan_cache_test.go:1876]: should check “0”？

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: pkg/planner/core/plan_cache_test.go (hunk 1 above)
Comment (qw4990) [reply pkg/planner/core/plan_cache_test.go]: Thanks, it's updated.

**Review by AilinKid — APPROVED** [review]

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**